library/common/main_interface.h

```diff
@@ -1,12 +1,214 @@
 #pragma once
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
 
 // NOLINT(namespace-envoy)
 
+/**
+ * Handle to an Envoy engine instance. Valid only for the lifetime of the engine and not intended
+ * for any external interpretation or use.
+ */
+typedef uint64_t envoy_engine_t;
+
+/**
+ * Handle to an outstanding Envoy HTTP stream. Valid only for the duration of the stream and not
+ * intended for any external interpretation or use.
+ */
+typedef uint64_t envoy_stream_t;
+
+/**
+ * Result codes returned by all calls made to this interface.
+ */
+typedef enum { ENVOY_SUCCESS, ENVOY_FAILURE } envoy_status_t;
+
+/**
+ * Error code associated with terminal status of a HTTP stream or request.
+ */
+typedef enum {} envoy_error_code_t;
+
+/**
+ * Common abstraction for strings.
+ */
+typedef struct {
+  uint64_t length;
+  char* data;
+} envoy_string;
+
+/**
+ * Holds data about an HTTP request.
+ */
+typedef struct {
+  // Status of the Envoy HTTP request. Note that the request might have failed inline.
+  // Thus the status should be checked before pursuing other operations on the request.
+  envoy_status_t status;
+  // Handle to the Envoy HTTP request.
+  envoy_request_t request;
+} envoy_request;
+
+/**
+ * Holds data about an HTTP stream.
+ */
+typedef struct {
+  // Status of the Envoy HTTP stream. Note that the stream might have failed inline.
+  // Thus the status should be checked before pursuing other operations on the stream.
+  envoy_status_t status;
+  // Handle to the Envoy HTTP Stream.
+  envoy_stream_t stream;
+} envoy_stream;
+
+/**
+ * Holds a single name/value header.
+ */
+typedef struct {
+  envoy_string name;
+  // Multiple header values for the same header name are supported via a comma-delimited string.
+  envoy_string value;
+} envoy_header;
+
+/**
+ * Holds an HTTP header map as an array of envoy_header structs.
+ */
+typedef struct {
+  // Size of the array.
+  uint64_t length;
+  // Array of headers.
+  envoy_header* headers;
+} envoy_headers;
```

Comment:
Out of curiosity, why do we need this type as opposed to using a hashmap?

Reply:
There are no maps in C. We could build that in this layer. But I'd rather use this simple struct and build better data structures in the c++ implementation of this header.